retry.go

```diff
@@ -16,53 +16,101 @@ var DefaultRetryOption = RetryOption{
-	Timer         backoff.Timer
+	// NextBackOff is called after each retry to determine the next backoff duration.
+	NextBackOff func(ctx context.Context, re RetryEvent, nextBackOff time.Duration) time.Duration
+
+	Backoff backoff.BackOff
+	Notify  backoff.Notify
+	Timer   backoff.Timer
+}
+
+// RetryEvent is the event fired when a retry happens
+type RetryEvent struct {
+	Attempt uint64
+	Since   time.Duration
+	Error   error
 }
 
 // retry constructs a do function with retry enabled according to the option.
 func (w *Workflow) retry(opt *RetryOption) func(
 	ctx context.Context,
-	fn func(context.Context) error,
+	do func(context.Context) error,
 	notAfter time.Time, // the Step level timeout ddl
 ) error {
 	if opt == nil {
-		return func(ctx context.Context, fn func(context.Context) error, notAfter time.Time) error {
-			return fn(ctx)
-		}
+		return func(ctx context.Context, do func(context.Context) error, notAfter time.Time) error { return do(ctx) }
 	}
-	return func(ctx context.Context, fn func(context.Context) error, notAfter time.Time) error {
+	return func(ctx context.Context, do func(context.Context) error, notAfter time.Time) error {
 		backOff := opt.Backoff
 		backOff = backoff.WithContext(backOff, ctx)
+		if !notAfter.IsZero() {
+			backOff = &backOffStopIfTimeout{BackOff: backOff, NotAfter: notAfter, Now: w.Clock.Now}
+		}
 		if opt.Attempts > 0 {
 			backOff = backoff.WithMaxRetries(backOff, opt.Attempts-1)
 		}
-		attempt := uint64(0)
+		retried := func(ctx context.Context, e RetryEvent) {}
+		if opt.NextBackOff != nil {
+			b := &backOffWithEvent{BackOff: backOff, nextBackOff: opt.NextBackOff}
+			retried = b.retried
+			backOff = b
+		}
+		e := RetryEvent{Attempt: 0}
 		start := w.Clock.Now()
 		return backoff.RetryNotifyWithTimer(
 			func() error {
-				defer func() { attempt++ }()
+				defer func() {
+					retried(ctx, e)
+					e.Attempt++
+				}()
+				ctxPerTry := ctx
 				if opt.TimeoutPerTry > 0 {
-					var cancel func()
-					ctx, cancel = w.Clock.WithTimeout(ctx, opt.TimeoutPerTry)
+					var cancel context.CancelFunc
+					ctxPerTry, cancel = w.Clock.WithTimeout(ctx, opt.TimeoutPerTry)
 					defer cancel()
 				}
-				if err := fn(ctx); err != nil {
-					switch {
-					case !notAfter.IsZero() && w.Clock.Now().After(notAfter), // Step level timeout
-						opt.StopIf != nil && opt.StopIf(ctx, attempt, w.Clock.Since(start), err):
-						return backoff.Permanent(err)
-					default:
-						return err
-					}
-				}
-				return nil
+				err := do(ctxPerTry)
+				e.Since = w.Clock.Since(start)
+				e.Error = err
+				return err
 			},
 			backOff,
 			opt.Notify,
 			opt.Timer,
 		)
 	}
 }
+
+type backOffWithEvent struct {
+	backoff.BackOff
+	nextBackOff func(context.Context, RetryEvent, time.Duration) time.Duration
+
+	ctx context.Context
+	e   RetryEvent
+}
+
+func (b *backOffWithEvent) NextBackOff() time.Duration {
+	bkof := b.BackOff.NextBackOff()
+	if b.nextBackOff == nil || bkof == backoff.Stop {
```

Comment:
Said, the original backoff stop has higher priority than the nextBackOff. Would user always understand it in that way?

Reply:
the orignal (inner) backoff like "max attempts", or "timeout" fires `Stop`, then the retry is stopped without calling `NextBackOff` callback

Reply:
You're right, we need to clarify this behavior

Reply:
Yeah, either logic is okay. As a good framework, we'd better design it less ambiguity.